src/test/scala/com/databricks/spark/redshift/RedshiftSourceSuite.scala

```diff
@@ -0,0 +1,486 @@
+  test("DefaultSource can load Redshift UNLOAD output to a DataFrame") {
+
+    val params = Map("url" -> "jdbc:postgresql://foo/bar",
+      "tempdir" -> "tmp",
+      "dbtable" -> "test_table",
+      "aws_access_key_id" -> "test1",
+      "aws_secret_access_key" -> "test2")
+
+    val jdbcWrapper = prepareUnloadTest(params)
+    val testSqlContext = new SQLContext(sc)
+
+    // Assert that we've loaded and converted all data in the test file
+    val source = new DefaultSource(jdbcWrapper)
+    val relation = source.createRelation(testSqlContext, params)
+    val df = testSqlContext.baseRelationToDataFrame(relation)
+
+    df.rdd.collect() zip expectedData foreach {
+      case (loaded, expected) =>
+        loaded shouldBe expected
+    }
+  }
+
+  test("DefaultSource supports simple column filtering") {
+
+    val params = Map("url" -> "jdbc:postgresql://foo/bar",
+      "tempdir" -> "tmp",
+      "dbtable" -> "test_table",
+      "aws_access_key_id" -> "test1",
+      "aws_secret_access_key" -> "test2")
+
+    val jdbcWrapper = prepareUnloadTest(params)
+    val testSqlContext = new SQLContext(sc)
+
+    // Construct the source with a custom schema
+    val source = new DefaultSource(jdbcWrapper)
+    val relation = source.createRelation(testSqlContext, params, TestUtils.testSchema)
+
+    val rdd = relation.asInstanceOf[PrunedFilteredScan].buildScan(Array("testByte", "testBool"), Array.empty[Filter])
+    val prunedExpectedValues =
+      Array(Row(1.toByte, true),
+            Row(1.toByte, false),
+            Row(0.toByte, null),
+            Row(0.toByte, false),
+            Row(null, null))
+
+    rdd.collect() zip prunedExpectedValues foreach {
+      case (loaded, expected) => loaded shouldBe expected
+    }
+  }
+
+  test("DefaultSource supports user schema, pruned and filtered scans") {
+
+    val params = Map("url" -> "jdbc:postgresql://foo/bar",
+      "tempdir" -> "tmp",
+      "dbtable" -> "test_table",
+      "aws_access_key_id" -> "test1",
+      "aws_secret_access_key" -> "test2")
+
+    val jdbcWrapper = prepareUnloadTest(params)
+    val testSqlContext = new SQLContext(sc)
+
+    // Construct the source with a custom schema
+    val source = new DefaultSource(jdbcWrapper)
+    val relation = source.createRelation(testSqlContext, params, TestUtils.testSchema)
+
+    // Define a simple filter to only include a subset of rows
+    val filters: Array[Filter] =
+      Array(EqualTo("testBool", true),
+            EqualTo("testString", "Unicode是樂趣"),
+            GreaterThan("testDouble", 1000.0),
+            LessThan("testDouble", Double.MaxValue),
+            GreaterThanOrEqual("testFloat", 1.0f),
+            LessThanOrEqual("testInt", 43))
+    val rdd = relation.asInstanceOf[PrunedFilteredScan].buildScan(Array("testByte", "testBool"), filters)
+
+    // We should now only have one matching row, with two columns
+    val filteredExpectedValues = Array(Row(1, true))
+    rdd.collect() zip filteredExpectedValues foreach {
```

Comment:
The `zip` here is funny given that this test only expects to receive one row as a correct answer.